### DISCUSSION TITLE
Custom runtimeclass App build fail

### DISCUSSION THREAD

**Discussion by EmmettTsai** [microsoft/WindowsAppSDK#1936] [Q&A] [answered]
Hi,

I want to add some public function to App struct, and call from some page.
But I can't build my code success, and got the error message:
`cannot convert argument 1 from 'winrt::Microsoft::UI::Xaml::Application' to 'std::nullptr_t'`

I reduce my project to a simple sample code.
I add a runtimeclass to App.idl and add a GetValue function:

```
//App.idl
namespace App1
{
    [default_interface]
    runtimeclass App : Microsoft.UI.Xaml.Application
    {
        App();
        Int32 GetValue();
    }
}
```

```
//App.xaml.h
#include "App.xaml.g.h"
#include "App.g.h"

namespace winrt::App1::implementation
{
    struct App : AppT<App>
    {
        App();

        void OnLaunched(Microsoft::UI::Xaml::LaunchActivatedEventArgs const&);
        int GetValue();

    private:
        winrt::Microsoft::UI::Xaml::Window window{ nullptr };
        int value;
    };
}

namespace winrt::App1::factory_implementation
{
    struct App : AppT<App, implementation::App>
    {
    };
}
```

and call GetValue in MainWindow

```
//MainWindow.xaml.cpp
    MainWindow::MainWindow()
    {
        InitializeComponent();

        auto app = Microsoft::UI::Xaml::Application::Current().as<App1::App>();
        int value = app.GetValue();
    }
```

build error message:
```
1>E:\WinUI3\App1\App1\Generated Files\App.g.cpp(10,12): error C2664: 'winrt::App1::App::App(std::nullptr_t) noexcept': cannot convert argument 1 from 'winrt::Microsoft::UI::Xaml::Application' to 'std::nullptr_t'
1>E:\WinUI3\App1\App1\Generated Files\App.g.cpp(10,44): message : No user-defined-conversion operator available that can perform this conversion, or the operator cannot be called
1>E:\WinUI3\App1\App1\Generated Files\winrt\impl\App1.2.h(15,9): message : see declaration of 'winrt::App1::App::App'
```

I can't find any related sample code.
Does WinUI3 support custom application struct?
Or is there something wrong with my code?
Thansk.

NuGet package version
1.0.0

Packaging type
Packaged (MSIX)

Windows version
Windows 10 version 20H1 (19043.1415)

IDE
Visual Studio 2019

Additional context
Project source:
[App1.zip](https://github.com/microsoft/WindowsAppSDK/files/7783368/App1.zip)

**Comment by jaigak** [comment, marked as answer]
Have you included App.h in MainWindow.xaml.h or MainWindow.xaml.cpp? Unlike other Windows Runtime classes, the App class shouldn't be defined in metadata (IDL) as its internal to the project. Instead, use the get_self method to get a pointer to the instance of the implementation type and access it that way.

**Reply by EmmettTsai** [reply]
Yes, I have include App.xaml.h in MainWindow.xaml.cpp, but still build fail.

After remove App runtime class in App.idl,
and remove winrt::App1::factory_implementation in App.xaml.h,
my project can build success and work fine.

Thank you very much.
